// @liquescens/auth-js/OAuth2.js
const Defaults =
{
    Google:
    {
        auth_uri: 'https://accounts.google.com/o/oauth2/v2/auth',
        scope: 'openid email'
    },
    GitHub:
    {
        auth_uri: `https://github.com/login/oauth/authorize`,
        scope: 'user'
    },
    Microsoft:
    {
        auth_uri: `https://login.microsoftonline.com/${'common'}/oauth2/v2.0/authorize`,
        scope: 'openid https://graph.microsoft.com/.default'
    },
    LinkedIn:
    {
        auth_uri: 'https://www.linkedin.com/oauth/v2/authorization',
        scope: 'profile email'
    }
};

export class OAuth2
{
    static Defaults = Defaults;

    /**
     * @param {import("./index.js").OAuth2Configuration} configuration
     */
    constructor(configuration)
    {
        this.configuration = configuration;
    }

    /**
     * @param {string} redirect_uri 
     */
    authenticate(redirect_uri)
    {
        const { id, client_id, scope } = this.configuration;
        const state = { provider: id };
        const parameters =
        {
            client_id,
            redirect_uri,
            response_type: 'code',
            scope,
            state: JSON.stringify(state)
        };
        window.location.href = `${this.configuration.auth_uri}?${new URLSearchParams(parameters)}`;
    }
}
